### PR TITLE
refactor(raft): move methods for reconfiguring out of RaftContext

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

This PR moves the methods for join,leave and anoint out of `RaftContext` to a new `ReconfigurationHelper`. Later force configure would be also added to this list.

This also moves `anoint` eventhough technically it is not a reconfiguring operation. However, it is still something that is triggered externaly and could fit in with the other operations.

## Related issues

This PR is in preparation to changes required for #16148

### CODE REVIEW

File: atomix/cluster/src/main/java/io/atomix/raft/impl/ReconfigurationHelper.java [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+
+public final class ReconfigurationHelper {
+
+  private final ThreadContext threadContext;
+  private final RaftContext raftContext;
+  private final Logger logger;
+
+  public ReconfigurationHelper(final RaftContext raftContext) {
```
Comment (npepinpe) [inline atomix/cluster/src/main/java/io/atomix/raft/impl/ReconfigurationHelper.java:46]: :wrench: If you plan on unit testing this, then I would not depend directly on the Raft context, it's too complex to stub or mock.

Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/impl/ReconfigurationHelper.java:46]: I tried to not directly use it :sob:. But it is too dependent on `RaftContext`. It need access to runtime info - like who is the leader, trigger transition etc.

File: atomix/cluster/src/main/java/io/atomix/raft/impl/ReconfigurationHelper.java
Comment (npepinpe) [inline atomix/cluster/src/main/java/io/atomix/raft/impl/ReconfigurationHelper.java:1]: :upside_down_face: I find huge inline lambdas hard to read and more error prone (as they can close on more state), but it's a very minor thing

File: atomix/cluster/src/main/java/io/atomix/raft/impl/DefaultRaftServer.java [hunk 2]
```diff
@@ -107,12 +107,12 @@ public CompletableFuture<RaftServer> join(final Collection<MemberId> cluster) {
 
   @Override
   public CompletableFuture<RaftServer> leave() {
-    return context.leave().thenApply(v -> this);
+    return new ReconfigurationHelper(context).leave().thenApply(v -> this);
```
Comment (npepinpe) [inline atomix/cluster/src/main/java/io/atomix/raft/impl/DefaultRaftServer.java:110]: :upside_down_face: Can probably be created only once? :shrug:

**Review by npepinpe — APPROVED** [review]
:rocket:

File: atomix/cluster/src/main/java/io/atomix/raft/impl/DefaultRaftServer.java (hunk 2 above)
Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/impl/DefaultRaftServer.java:110]: There is no need to keep a long living object, right? These operations are not expected to be called frequently. join/leave will be called atmost once during the server's lifetime.

File: atomix/cluster/src/main/java/io/atomix/raft/impl/ReconfigurationHelper.java
Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/impl/ReconfigurationHelper.java:1]: Refactored them.

**Pull request merged**